feat: Advanced Fluid Interaction

null2264 opened the pull request:
This PR not only introduced "advanced custom generator", but also reworked how the mod handles fluid interaction as a whole to make it more compatible with most mods that control fluid interaction since it will only cancel the original function on success and doesn't directly touch how Minecraft handle fluid interactions.

Planned to be released for v4.0.

## Features

### Advanced Generators
This will allow players (and modpack makers) to create a custom generator with custom fluids or blocks
```json5
{
    // You can disable advanced generators by setting it like this:
    // advanced: {}
    //
    // Or just delete it entirely.

    "advanced": {
        // This is the fluid that will touch another fluid (in this example, it's lava touching water)
        "minecraft:lava": {
            "minecraft:water": {  // For this example, we mimic cobblestone/stone generator's behaviour
                // Result format is similar to custom generators' format
                "results": {  // Mimic cobblestone generator
                    "*": [  // Unlike custom generators' format, wildcard is supported as a modifier (accept any block)
                        {
                            "id": "minecraft:cobblestone",
                            "weight": 100.0
                        }
                    ]
                },
                // Only triggered when lava touch water from the top
                "resultsFromTop": {  // Mimic stone generator, only works for fluids
                    "*": [
                        {
                            "id": "minecraft:stone",
                            "weight": 100.0
                        }
                    ]
                }
            },
            "b:minecraft:blue_ice": {  // Mimic basalt generator's behaviour
                "results": {  // Blocks only support "results"
                    "minecraft:soul_soil": [
                        {
                            "id": "minecraft:basalt",
                            "weight": 100.0
                        }
                    ]
                }
            },
            "create:honey": {
                "silent": true,  // This will prevent generator from making the extinguish sound effect when block is generated
                "results": {
                    "*": [
                        {
                            "id": "create:limestone",
                            "weight": 100.0
                        }
                    ]
                }
            }
        }
    }
}
```

### Fluid Interaction API
This is mainly for mod developers. With this developers can now create a plugin for CobbleGen, and register a fluid interaction more easily.

To make one, you need to create a class that implements `CobbleGenPlugin`, for example:
```java
public class CreateFluidInteraction implements CobbleGenPlugin
{
    public void registerInteraction() {
        if (!FabricLoader.getInstance().isModLoaded("create")) return;
        FLUID_INTERACTION.addGenerator(
                Fluids.LAVA,
                new CobbleGenerator(List.of(new WeightedBlock("create:limestone", 1.0)),
                                    AllFluids.HONEY.get(),
                                    false
                )
        );
        FLUID_INTERACTION.addGenerator(
                Fluids.LAVA,
                new CobbleGenerator(List.of(new WeightedBlock("create:scoria", 1.0)),
                                    AllFluids.CHOCOLATE.get(),
                                    false
                )
        );
    }
}
```

Then add your plugin location to a `cobblegen_plugin` entrypoint in `fabric.mod.json`, for example:
```json
  "entrypoints": {
    "cobblegen_plugin": [
      "io.github.null2264.cobblegen.integration.create.CreateFluidInteraction"
    ]
  },
```

## Checklist
- [x] Fluid Interaction API  
  This should make CobbleGen more compatible with most mods that control fluid interactions
  - [x] Make other generators use this API
  - [x] Implement Advanced Custom Generator
- [x] Testing
  - [x] Without Create/Porting Lib
    - [x] 1.18.2
    - [x] 1.19.2
  - [x] With Create/Porting Lib
    - [x] 1.18.2
    - [x] 1.19.2

null2264 committed:
feat: Advanced fluid interaction

null2264 committed:
feat: Wildcard support

null2264 committed:
chore: Exit early if advanced generator is not configured

null2264 committed:
chore: Disable advanced generators by default

null2264 committed:
feat: Allow users to define advanced custom stone generators

null2264 committed:
chore: Simplify code

null2264 committed:
chore: Simplify code

null2264 committed:
feat: Silent generator

null2264 committed:
merge: Merge branch '1.18+' into feat/fluidInteraction

DemonSlayer112 commented:
This looks amazing! :D
I assume if we wanted a basalt gen to work with only 1 block, like soul sand only, we would specifiy "blue_ice" as being "air"?

And if we did do this, how would the mod treat a normal basalt gen(lava near soulsoil and ice) apart from lava(near just....a soulsoil and air for instance)?

also for a little fun for people air is a outputable block xD
Ig this is amazing for farm testing

null2264 commented:
>  I assume if we wanted a basalt gen to work with only 1 block, like soul sand only, we would specifiy "blue_ice" as being "air"?

Correct

> And if we did do this, how would the mod treat a normal basalt gen(lava near soulsoil and ice) apart from lava(near just....a soulsoil and air for instance)?

Because of how the mod is coded, it'll probably overwrite normal basalt gen unless you surround the generated block with blocks other than air and a blue ice.

Here's a working build for testing.
[cobblegen.zip](https://github.com/null2264/CobbleGen/files/11327787/cobblegen.zip)

null2264 committed:
refactor: Merge generators

null2264 committed:
chore: Optimize code

null2264 committed:
chore: Improving readability

null2264 committed:
chore: Simplify code

null2264 committed:
feat: Fluid Interaction API (Big overhaul)
REI/EMI/JEI is broken in this commit, because config handler is overhauled to be processed on launch (and `/reload`). This will allow other people to add their own generator with relative ease.

null2264 committed:
fix: REI/EMI/JEI integration

null2264 committed:
fix: Pipe always generate cobblestone instead of stone

null2264 committed:
feat: Built-in plugin for Create mod

null2264 committed:
refactor: Reformat code
